add wasm support (#228)
subquery/documentation: docs/build/substrate-wasm.md

# Substrate WASM Support

We provide a custom data source processor for [Substrate WASM contract](https://ink.substrate.io/why-webassembly-for-smart-contracts). This offers a simple way to filter and index both WASM and Substrate activity on many Polkadot networks within a single SubQuery project.

**Tested and Supported networks**

| Network Name   | Websocket Endpoint                         | Dictionary Endpoint                                           |
|----------------|--------------------------------------------|---------------------------------------------------------------|
| Shiden         | `wss://shiden.api.onfinality.io/public-ws` | `https://api.subquery.network/sq/subquery/shiden-dictionary`  |
| Shibuya        |                                            | `https://api.subquery.network/sq/subquery/shibuya-dictionary` |

> 

Reviewer: Any other network this could support in theory?

Author: Added one more network: Edgeware

Reviewer: Thanks

| filter | [Call Filter](substrate-wasm.md#call-filters) | No       | Filter the data source to execute           |

### Call Filters

| Field    | Type   | Example(s)                                   | Description                                                                         |
|----------| ------ |----------------------------------------------|-------------------------------------------------------------------------------------|
| selector | String | `0x681266a0`                                 | [Selector](https://ink.substrate.io/macros-attributes/selector/) identifies a method |
| method   | String | `approve`                                    | The label for the message, same as the function called on the contract              |
| from     | String | `0x6bd193ee6d2104f14f94e2ca6efefae561a4334b` | The sender, a Substrate Account that submitted the transaction                      |

### Handler Functions

Unlike a normal handler you will not get a `SubstrateExtrinsic` as the parameter, instead you will use `substrate/WasmCall` and receive `WasmCall` as the parameter. 

> The `WasmCall` includes these useful information:

Reviewer: ```suggestion
The `WasmCall` includes following data:
```

| kind   | `substrate/WasmEvent`          | Yes      | Specifies that this is an Event type handler |
| filter | [Event Filter](#event-filters) | No       | Filter the data source to execute            |

### Event Filters

| Field      | Type      | Example(s)                                          | Description                                                          |
|------------|-----------|-----------------------------------------------------|----------------------------------------------------------------------|
| from       | String    | `YeuN6quBhpFnd5DyWNCrwGvpyBgq51Q3nbVMSsQJ6toPXSf`   | The sender of the contract call which trigger this contract event    |
| contract   | String    | `a6Yrf6jAPUwjoi5YvvoTE4ES5vYAMpV55ZCsFHtwMFPDx7H`   | A contract address where the event is from                           |
| identifier | String    | `Transfer`                                          | The label of the event, it identifies the event type in the contract |

### Handler Functions

Unlike a normal handler you will not get a `SubstrateEvent` as the parameter, instead you will get a `substrate/WasmEvent` with `WasmEvent` as its parameter.

> 

Reviewer: ```suggestion
The `WasmEvent` includes useful information:
```

    }else{
        logger.info(`Decode call failed ${call.hash}`)
    }
    approval.contractAddress = call.dest.toString();
    await approval.save();
}
```

</CodeGroupItem>
</CodeGroup>

## Data Source Example

This is an extract from the `project.yaml` manifest file.

<CodeGroup>
<CodeGroupItem title="Frontier WASM" active>

```yaml
dataSources:
  - kind: substrate/Wasm
    startBlock: 970733
    processor:
      file: ./node_modules/@subql/substrate-wasm-processor/dist/bundle.js
      options:
        abi: erc20
        contract: 'a6Yrf6jAPUwjoi5YvvoTE4ES5vYAMpV55ZCsFHtwMFPDx7H'
    assets:
      erc20:
        file: ./erc20Metadata.json
    mapping:
      file: ./dist/index.js
      handlers:
        - handler: handleSubstrateWasmEvent
          kind: substrate/WasmEvent
          filter:
            # from: 'xxxx'
            contract: 'a6Yrf6jAPUwjoi5YvvoTE4ES5vYAMpV55ZCsFHtwMFPDx7H'
            identifier: 'Transfer'
        - handler: handleSubstrateCall
          kind: substrate/WasmCall
          filter:
            selector: '0x681266a0'
            method: 'approve'
```

</CodeGroupItem>
</CodeGroup>

## Querying contracts

> [SubstrateWasmProvider](https://github.com/subquery/datasource-processors/blob/main/packages/substrate-wasm/src/substrateWasmProvider.ts) provide contract query method like `getStorage` and `call`.

Reviewer: ```suggestion
[SubstrateWasmProvider](https://github.com/subquery/datasource-processors/blob/main/packages/substrate-wasm/src/substrateWasmProvider.ts) provides contract query method like `getStorage` and `call`.
```

## Known Limitations

> - Not support dynamic datasource at the moment, which means we are unable to import and handle contract dynamically.

Reviewer: ```suggestion
- Not supporting dynamic datasource at the moment. which means we are unable to import and handle contracts dynamically.
```

## Data Source Example

This is an extract from the `project.yaml` manifest file.

<CodeGroup>
<CodeGroupItem title="Frontier WASM" active>

```yaml
dataSources:
  - kind: substrate/Wasm
    startBlock: 970733
    processor:
      file: ./node_modules/@subql/substrate-wasm-processor/dist/bundle.js
      options:
        abi: erc20
        contract: 'a6Yrf6jAPUwjoi5YvvoTE4ES5vYAMpV55ZCsFHtwMFPDx7H'
    assets:
      erc20:
        file: ./erc20Metadata.json
    mapping:
      file: ./dist/index.js
      handlers:
        - handler: handleSubstrateWasmEvent
          kind: substrate/WasmEvent
          filter:
            # from: 'xxxx'
            contract: 'a6Yrf6jAPUwjoi5YvvoTE4ES5vYAMpV55ZCsFHtwMFPDx7H'
            identifier: 'Transfer'
        - handler: handleSubstrateCall
          kind: substrate/WasmCall
          filter:
            selector: '0x681266a0'
            method: 'approve'
```

</CodeGroupItem>
</CodeGroup>

## Querying contracts

[SubstrateWasmProvider](https://github.com/subquery/datasource-processors/blob/main/packages/substrate-wasm/src/substrateWasmProvider.ts) provides contract query method like `getStorage` and `call`.

## Known Limitations

> - Not supporting dynamic datasource at the moment. which means we are unable to import and handle contracts dynamically.

Reviewer: ```suggestion
- Not supporting dynamic datasource at the moment, which means we are unable to import and handle contracts dynamically.
```
```suggestion
- Not supporting dynamic datasource at the moment. which means we are unable to import and handle contracts dynamically.
```
